File: packages/py/analytics/src/bolsa_analytics/knowledge/dia_d_session_evidence.py
"""Evidence informe sesión C DÍA D — facts deterministas; LLM solo narra.

Bandas: favorable | mixed | adverse | incomplete.
Entrada: métricas auto/gated + gate (accept/reject) del sandbox D→hoy.
Salida: schema ``dia_d_session_evidence_v1`` (claims, warnings, paragraphs×3).

No FA, no Coach, no Belief, no DEMO live.
API narración: ``POST /api/ai/dia-d/session-evidence``.
Persist Fase 2: ``POST /api/research/dia-d-session-evidence`` (source=dia_d_session).
"""

from __future__ import annotations

from typing import Any, Literal

DiaDEvidenceBand = Literal["favorable", "mixed", "adverse", "incomplete"]

SCHEMA = "dia_d_session_evidence_v1"


def _fmt_pct(n: float) -> str:
    sign = "+" if n > 0 else ""
    return f"{sign}{n:.2f}%"


def resolve_band(payload: dict[str, Any]) -> DiaDEvidenceBand:
    mode = str(payload.get("mode") or "auto")
    auto = payload.get("auto") if isinstance(payload.get("auto"), dict) else {}
    gated = payload.get("gated") if isinstance(payload.get("gated"), dict) else {}
    gate = payload.get("gate") if isinstance(payload.get("gate"), dict) else {}
    accepted = int(gate.get("accepted") or 0)
    rejected = int(gate.get("rejected") or 0)
    auto_trades = int(auto.get("tradeCount") or 0)
    gated_ret = float(gated.get("totalReturnPct") or 0)
    auto_ret = float(auto.get("totalReturnPct") or 0)
    gated_dd = float(gated.get("maxDrawdownPct") or 0)
    auto_dd = float(auto.get("maxDrawdownPct") or 0)

    if mode != "auto" and accepted + rejected == 0 and auto_trades > 0:
        return "incomplete"
    delta = gated_ret - auto_ret
    if gated_ret <= -8 or delta <= -6:
        return "adverse"
    if gated_ret >= 0 and (delta >= -1 or rejected == 0) and gated_dd <= max(auto_dd, 15) + 2:
        return "favorable"
    if gated_ret >= 2 and delta >= 0:
        return "favorable"
    return "mixed"


def build_dia_d_session_evidence(payload: dict[str, Any]) -> dict[str, Any]:
    """Ensambla Evidence v1 desde métricas ya calculadas (sin FA/Coach)."""
    mode = str(payload.get("mode") or "auto")
    symbol = str(payload.get("symbol") or "?")
    strategy = str(payload.get("strategyLabel") or "#1")
    dia_d = str(payload.get("diaD") or "")
    end_date = str(payload.get("endDate") or "")
    auto = payload.get("auto") if isinstance(payload.get("auto"), dict) else {}
    gated = payload.get("gated") if isinstance(payload.get("gated"), dict) else {}
    gate = payload.get("gate") if isinstance(payload.get("gate"), dict) else {}
    accepted = int(gate.get("accepted") or 0)
    rejected = int(gate.get("rejected") or 0)
    gated_ret = float(gated.get("totalReturnPct") or 0)
    auto_ret = float(auto.get("totalReturnPct") or 0)
    gated_dd = float(gated.get("maxDrawdownPct") or 0)
    gated_ops = int(gated.get("tradeCount") or 0)
    auto_ops = int(auto.get("tradeCount") or 0)
    final_eq = float(gated.get("finalEquity") or 0)
    delta = gated_ret - auto_ret
    band = resolve_band(payload)

    claims = [
        f"Modo {mode} · {symbol} · #1 {strategy}",
        f"Ventana {dia_d} → {end_date}",
        f"Retorno sesión {_fmt_pct(gated_ret)} · DD {_fmt_pct(gated_dd)} · {gated_ops} ops",
    ]
    if mode != "auto":
        claims.append(
            f"Vs Auto: retorno {_fmt_pct(delta)} · ops {gated_ops}/{auto_ops} · "
            f"gate OK/KO {accepted}/{rejected}"
        )
    else:
        claims.append(f"Trayectoria Auto (#1 congelada) · capital fin {final_eq:.2f}")
    if band == "incomplete":
        claims.append("Aún no hay decisiones de gate: el informe es provisional.")

    warnings: list[str] = []
    if band == "incomplete":
        warnings.append("Informe incompleto: sin decisiones Semi/Manual aún.")
    if gated_ops == 0 and mode == "auto" and auto_ops == 0:
        warnings.append("El run Auto no generó operaciones en D→hoy.")
    if gated_dd >= 20:
        warnings.append("Max DD ≥ 20%: revisar tamaño y filtros.")

    if band == "incomplete":
        p1 = (
            f"Sesión DÍA D de {symbol} en modo {mode}: hay {auto_ops} señales Auto "
            "pendientes de Aceptar/Rechazar. Hasta decidir, el equity gated no refleja tu criterio."
        )
    elif band == "favorable":
        p1 = (
            f"En {dia_d}→{end_date}, la sesión {mode} de {symbol} cierra con retorno "
            f"{_fmt_pct(gated_ret)} (DD {_fmt_pct(gated_dd)}). Coherente o mejor que el Auto."
        )
    elif band == "adverse":
        p1 = (
            f"La sesión {mode} de {symbol} termina con retorno {_fmt_pct(gated_ret)} y DD "
            f"{_fmt_pct(gated_dd)}. Respecto al Auto ({_fmt_pct(auto_ret)}) el delta es {_fmt_pct(delta)}."
        )
    else:
        p1 = (
            f"Resultado mixto en {symbol}: retorno {_fmt_pct(gated_ret)} vs Auto "
            f"{_fmt_pct(auto_ret)} (delta {_fmt_pct(delta)}), con {gated_ops} ops."
        )

    if mode == "auto":
        p2 = (
            "Modo Auto ejecuta todas las señales de #1 sin filtro humano. "
            "Úsalo como referencia; Semi/Manual miden el valor de tus vetos."
        )
    elif rejected > 0:
        p2 = (
            f"Gate: {accepted} aceptadas y {rejected} rechazadas. "
            "Un buy rechazado anula su sell; el equity se reescribe solo con accepts."
        )
    else:
        p2 = (
            f"Gate: {accepted} aceptadas y ninguna rechazo. "
            "La trayectoria se acerca al Auto; conviene contrastar DD y nº de ops."
        )

    if band == "incomplete":
        p3 = (
            "Siguiente: recorre la película, decide cada señal y vuelve a este informe. "
            "No es consejo de inversión ni despliegue DEMO."
        )
    elif band == "adverse":
        p3 = (
            "Revisa si los rechazos evitaron pérdidas o cortaron winners. "
            "Contrasta con el embudo ≤ D. Sandbox ≠ DEMO live."
        )
    else:
        p3 = (
            "Guarda el aprendizaje: qué señales vetaste y por qué. "
            "El embudo ≤ D es otro informe; este solo cubre D→hoy. Sandbox ≠ DEMO live."
        )

    confidence: Literal["HIGH", "MEDIUM", "LOW"]
    if band == "incomplete":
        confidence = "LOW"
    elif mode == "auto" or accepted + rejected >= 3:
        confidence = "HIGH"
    else:
        confidence = "MEDIUM"

    return {
        "schemaVersion": SCHEMA,
        "band": band,
        "confidence": confidence,
        "claims": claims,
        "warnings": warnings,
        "metrics": {
            "mode": mode,
            "returnPct": gated_ret,
            "maxDrawdownPct": gated_dd,
            "tradeCount": gated_ops,
            "finalEquity": final_eq,
            "autoReturnPct": auto_ret,
            "returnDeltaVsAutoPct": delta,
            "accepted": accepted,
            "rejected": rejected,
        },
        "paragraphs": [p1, p2, p3],
        "disclaimer": (
            "Interpretación Evidence de la sesión sandbox DÍA D. "
            "No recalcula FA ni Coach. No es consejo de inversión ni escribe la DEMO live."
        ),
    }


def evidence_prompt_variables(evidence: dict[str, Any], payload: dict[str, Any]) -> dict[str, str]:
    claims = evidence.get("claims") if isinstance(evidence.get("claims"), list) else []
    warnings = evidence.get("warnings") if isinstance(evidence.get("warnings"), list) else []
    paragraphs = evidence.get("paragraphs") if isinstance(evidence.get("paragraphs"), list) else []
    return {
        "symbol": str(payload.get("symbol") or ""),
        "mode": str(payload.get("mode") or ""),
        "band": str(evidence.get("band") or ""),
        "confidence": str(evidence.get("confidence") or ""),
        "claims": " | ".join(str(c) for c in claims),
        "warnings": " | ".join(str(w) for w in warnings) or "(ninguno)",
        "heuristic": " || ".join(str(p) for p in paragraphs),
        "disclaimer": str(evidence.get("disclaimer") or ""),
    }

File: apps/web/src/features/trading/dia-d-session-evidence.ts
/**
 * Informe Evidence sesión C (DÍA D) — solo interpreta métricas ya calculadas.
 * No recalcula FA ni Coach. Sandbox ≠ DEMO.
 *
 * Paridad con Python ``bolsa_analytics.knowledge.dia_d_session_evidence``.
 * UI: `trading-dia-d-replay-panel.tsx` · API narración: `explainDiaDSessionEvidence`.
 *
 * @see docs/engineering/backtesting-dia-d-premises-2026-07-31.md
 */

export type DiaDEvidenceBand = 'favorable' | 'mixed' | 'adverse' | 'incomplete';

export type DiaDSessionEvidenceInput = {
  mode: 'manual' | 'semi' | 'auto';
  symbol: string;
  strategyLabel: string;
  diaD: string;
  endDate: string;
  initialCash: number;
  auto: {
    totalReturnPct: number;
    maxDrawdownPct: number;
    tradeCount: number;
    finalEquity: number;
  };
  /** Trayectoria tras gate (en Auto = igual que auto). */
  gated: {
    totalReturnPct: number;
    maxDrawdownPct: number;
    tradeCount: number;
    finalEquity: number;
  };
  gate: {
    accepted: number;
    rejected: number;
  };
};

export type DiaDSessionEvidenceV1 = {
  schemaVersion: 'dia_d_session_evidence_v1';
  band: DiaDEvidenceBand;
  confidence: 'HIGH' | 'MEDIUM' | 'LOW';
  claims: string[];
  warnings: string[];
  metrics: {
    mode: string;
    returnPct: number;
    maxDrawdownPct: number;
    tradeCount: number;
    finalEquity: number;
    autoReturnPct: number;
    returnDeltaVsAutoPct: number;
    accepted: number;
    rejected: number;
  };
  paragraphs: [string, string, string];
  disclaimer: string;
};

function fmtPct(n: number): string {
  const sign = n > 0 ? '+' : '';
  return `${sign}${n.toFixed(2)}%`;
}

function resolveBand(input: DiaDSessionEvidenceInput): DiaDEvidenceBand {
  const { mode, auto, gated, gate } = input;
  if (mode !== 'auto' && gate.accepted + gate.rejected === 0 && auto.tradeCount > 0) {
    return 'incomplete';
  }
  const delta = gated.totalReturnPct - auto.totalReturnPct;
  if (gated.totalReturnPct <= -8 || delta <= -6) return 'adverse';
  if (
    gated.totalReturnPct >= 0 &&
    (delta >= -1 || gate.rejected === 0) &&
    gated.maxDrawdownPct <= Math.max(auto.maxDrawdownPct, 15) + 2
  ) {
    return 'favorable';
  }
  if (gated.totalReturnPct >= 2 && delta >= 0) return 'favorable';
  return 'mixed';
}

function buildClaims(input: DiaDSessionEvidenceInput, band: DiaDEvidenceBand): string[] {
  const delta = input.gated.totalReturnPct - input.auto.totalReturnPct;
  const claims: string[] = [
    `Modo ${input.mode} · ${input.symbol} · #1 ${input.strategyLabel}`,
    `Ventana ${input.diaD} → ${input.endDate}`,
    `Retorno sesión ${fmtPct(input.gated.totalReturnPct)} · DD ${fmtPct(input.gated.maxDrawdownPct)} · ${input.gated.tradeCount} ops`,
  ];
  if (input.mode !== 'auto') {
    claims.push(
      `Vs Auto: retorno ${fmtPct(delta)} · ops ${input.gated.tradeCount}/${input.auto.tradeCount} · gate OK/KO ${input.gate.accepted}/${input.gate.rejected}`,
    );
  } else {
    claims.push(`Trayectoria Auto (#1 congelada) · capital fin ${input.gated.finalEquity.toFixed(2)}`);
  }
  if (band === 'incomplete') {
    claims.push('Aún no hay decisiones de gate: el informe es provisional.');
  }
  return claims;
}

function buildParagraphs(
  input: DiaDSessionEvidenceInput,
  band: DiaDEvidenceBand,
): [string, string, string] {
  const delta = input.gated.totalReturnPct - input.auto.totalReturnPct;
  const p1 =
    band === 'incomplete'
      ? `Sesión DÍA D de ${input.symbol} en modo ${input.mode}: hay ${input.auto.tradeCount} señales Auto pendientes de Aceptar/Rechazar. Hasta decidir, el equity gated no refleja tu criterio.`
      : band === 'favorable'
        ? `En ${input.diaD}→${input.endDate}, la sesión ${input.mode} de ${input.symbol} cierra con retorno ${fmtPct(input.gated.totalReturnPct)} (DD ${fmtPct(input.gated.maxDrawdownPct)}). El resultado es coherente o mejor que el Auto de referencia.`
        : band === 'adverse'
          ? `La sesión ${input.mode} de ${input.symbol} termina con retorno ${fmtPct(input.gated.totalReturnPct)} y DD ${fmtPct(input.gated.maxDrawdownPct)}. Respecto al Auto (${fmtPct(input.auto.totalReturnPct)}) el delta es ${fmtPct(delta)}.`
          : `Resultado mixto en ${input.symbol}: retorno ${fmtPct(input.gated.totalReturnPct)} vs Auto ${fmtPct(input.auto.totalReturnPct)} (delta ${fmtPct(delta)}), con ${input.gated.tradeCount} ops ejecutadas.`;

  const p2 =
    input.mode === 'auto'
      ? `Modo Auto ejecuta todas las señales de #1 sin filtro humano. Úsalo como referencia; Semi/Manual miden el valor de tus vetos.`
      : input.gate.rejected > 0
        ? `Gate: ${input.gate.accepted} aceptadas y ${input.gate.rejected} rechazadas. Un buy rechazado anula su sell; el equity se reescribe solo con accepts.`
        : `Gate: ${input.gate.accepted} aceptadas y ninguna rechazo. La trayectoria se acerca al Auto; conviene contrastar DD y nº de ops.`;

  const p3 =
    band === 'incomplete'
      ? `Siguiente: recorre la película, decide cada señal y vuelve a este informe. No es consejo de inversión ni despliegue DEMO.`
      : band === 'adverse'
        ? `Revisa si los rechazos evitaron pérdidas o cortaron winners. Contrasta con el embudo ≤ D (no es el mismo informe). Sandbox ≠ DEMO live.`
        : `Guarda el aprendizaje: qué señales vetaste y por qué. El embudo ≤ D es otro informe; este solo cubre D→hoy. Sandbox ≠ DEMO live.`;

  return [p1, p2, p3];
}

export function buildDiaDSessionEvidence(
  input: DiaDSessionEvidenceInput,
): DiaDSessionEvidenceV1 {
  const band = resolveBand(input);
  const delta = input.gated.totalReturnPct - input.auto.totalReturnPct;
  const warnings: string[] = [];
  if (band === 'incomplete') {
    warnings.push('Informe incompleto: sin decisiones Semi/Manual aún.');
  }
  if (input.gated.tradeCount === 0 && input.mode === 'auto' && input.auto.tradeCount === 0) {
    warnings.push('El run Auto no generó operaciones en D→hoy.');
  }
  if (input.gated.maxDrawdownPct >= 20) {
    warnings.push('Max DD ≥ 20%: revisar tamaño y filtros.');
  }

  const confidence: DiaDSessionEvidenceV1['confidence'] =
    band === 'incomplete'
      ? 'LOW'
      : input.mode === 'auto' || input.gate.accepted + input.gate.rejected >= 3
        ? 'HIGH'
        : 'MEDIUM';

  return {
    schemaVersion: 'dia_d_session_evidence_v1',
    band,
    confidence,
    claims: buildClaims(input, band),
    warnings,
    metrics: {
      mode: input.mode,
      returnPct: input.gated.totalReturnPct,
      maxDrawdownPct: input.gated.maxDrawdownPct,
      tradeCount: input.gated.tradeCount,
      finalEquity: input.gated.finalEquity,
      autoReturnPct: input.auto.totalReturnPct,
      returnDeltaVsAutoPct: delta,
      accepted: input.gate.accepted,
      rejected: input.gate.rejected,
    },
    paragraphs: buildParagraphs(input, band),
    disclaimer:
      'Interpretación Evidence de la sesión sandbox DÍA D. No recalcula FA ni Coach. No es consejo de inversión ni escribe la DEMO live.',
  };
}

export const DIA_D_EVIDENCE_BAND_LABELS: Record<DiaDEvidenceBand, string> = {
  favorable: 'Favorable',
  mixed: 'Mixto',
  adverse: 'Adverso',
  incomplete: 'Incompleto',
};
